pkg/runtime/fake.go

```diff
@@ -147,6 +149,58 @@ type FakeRuntime struct {
 	FakeExecutor *FakeExecutor
 	FakeProfiler *FakeProfiler
 	FakeLock     *FakeLock
+	FakeSignal   *FakeSignal
+}
+
+// FakeSignal implements a fake signal handling for testing
+type FakeSignal struct {
+	lock    sync.Mutex
+	signals map[os.Signal]bool
+	channel chan os.Signal
+}
+
+// NewFakeSignal returns a FakeSignal
+func NewFakeSignal() *FakeSignal {
+	return &FakeSignal{
+		channel: make(chan os.Signal),
+		signals: map[os.Signal]bool{},
+	}
+}
+
+// Notify implements Signal's interface Notify method
+func (f *FakeSignal) Notify(signals ...os.Signal) <-chan os.Signal {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	for _, s := range signals {
+		f.signals[s] = true
+	}
+	return f.channel
+}
+
+// Reset implements Signal's interface Reset method
+func (f *FakeSignal) Reset(signals ...os.Signal) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	if len(signals) == 0 {
+		f.signals = map[os.Signal]bool{}
+	} else {
+		for _, s := range signals {
+			delete(f.signals, s)
+		}
+	}
+}
+
+// SendSignal sends the given signal to the signal notification channel if the signal was
+// previously specified in a call to Notify
+func (f *FakeSignal) SendSignal(signal os.Signal) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	if f.signals[signal] {
+		f.channel <- signal
+	}
```

Comment:
It's likely not the most important thing on a fake, but I would avoid attempting to send data to a channel while holding a mutex at the same time. If for some reason nobody is listening, we could hold the mutex indefinitely and unexpectedly block other operations.

A simple way to avoid this would be to release the mutex after checking `f.signals[signal]`, and then sending the signal into the channel.

Reply:
> if the signal is cleared from the notifications after the check, it will still be sent.

On production code, I think this should be fine. In the end, even if you guard pushing to the channel with the mutex, the receving code will not be guarded, so you could argue that the (receiving) code is reacting to the signal after you called `Reset`. Async/concurrent APIs need to draw this line at some point, and this one seems just as good as any other.

That being said...

> I'm wondering if I should just get rid of the map, which is used to prevent sending a signal to the channel if not explicitly asked for.

I'm perfectly okay with this. The only thing this map would be useful for would be to test that we are not subscribing to certain signals, which I don't see as something very useful to do. I'd vote to get rid of it entirely.